### PR TITLE
Add new completion methods n' flags to repo from edx-platform

### PR DESCRIPTION
**Pull request by yro**
### Description
Bring repo to parity with changes made in [PR 17310](https://github.com/edx/edx-platform/pull/17310).
Update code quality, test methods as needed.

### Testing
- [x] Unit tests

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review: @iloveagent57 

FYI: @edx/educator-neem

### Post-review
- [x] Rebase and squash commits

### CODE REVIEW

File: completion/models.py [hunk 1]
```diff
@@ -168,6 +168,38 @@ class BlockCompletion(TimeStampedModel, models.Model):
...
+            user=user,
+            course_key=course_key,
+        )
+        # will not return if <= 0.0
+        return {completion.block_key: completion.completion for completion in course_block_completions}
+
+    @classmethod
+    def get_latest_block_completed(cls, user, course_key):
+        """
+        query latest completion for course/user pair
```
Comment (iloveagent57) [inline completion/models.py]: You can squash this string all into one sentence: "Returns the most recently modified BlockCompletion object for the given user and course_key, or None if no such BlockCompletion exists."

File: completion/models.py [hunk 2]
```diff
@@ -168,6 +168,38 @@ class BlockCompletion(TimeStampedModel, models.Model):
 
     objects = BlockCompletionManager()
 
+    @classmethod
+    def get_course_completions(cls, user, course_key):
+        """
+        query all completions for course/user pair
```
Comment (iloveagent57) [inline completion/models.py]: This could be a little more descriptive: "Returns a dictionary mapping BlockKeys to completion values for all BlockCompletion records for the given user and course_key."

File: completion/tests/test_models.py [hunk 3]
```diff
@@ -142,3 +142,69 @@ def test_submit_batch_completion_without_waffle(self):
         with self.assertRaises(RuntimeError):
             blocks = [(self.block_key, 1.0)]
             models.BlockCompletion.objects.submit_batch_completion(self.user, self.course_key, blocks)
+
+
+class BatchCompletionMethodTests(CompletionSetUpMixin, TestCase):
```
Comment (iloveagent57) [inline completion/tests/test_models.py]: This class could be named better - it would be easy to confuse it for tests corresponding to the `submit_batch_completion()` method, rather than the class methods it actually tests.  `CompletionFetchingTestCase` maybe, or something else you can think of?

File: completion/tests/test_models.py [hunk 4]
```diff
@@ -142,3 +142,69 @@ def test_submit_batch_completion_without_waffle(self):
...
+    """
+    Test model methods:
+        latest completion per course, and all completions per course
+    """
+    COMPLETION_SWITCH_ENABLED = True
+
+    def setUp(self):
+        super(BatchCompletionMethodTests, self).setUp()
+        self.set_up_completion()
+        self.user = UserFactory.create()
```
Comment (iloveagent57) [inline completion/tests/test_models.py]: This comes for free from the `CompletionSetUpMixin` class.

Comment (iloveagent57) [reply completion/tests/test_models.py]: Although, maybe since you also have `self.other_user`, it's not a bad idea to just spend the extra line and re-define it (like you're doing).  So I guess just leave it.

File: completion/tests/test_models.py [hunk 5]
```diff
@@ -142,3 +142,69 @@ def test_submit_batch_completion_without_waffle(self):
...
+        for idx, block_key in enumerate(self.block_keys[0:3]):
+            models.BlockCompletion.objects.submit_completion(
+                user=self.user,
+                course_key=self.course_key,
+                block_key=block_key,
+                completion=1.0 - (0.2 * idx),
+            )
+
+        for idx, block_key in enumerate(self.block_keys[2:]):  # Wrong user
+
```
Comment (iloveagent57) [inline completion/tests/test_models.py]: you don't need this empty line

File: completion/models.py [hunk 6]
```diff
@@ -168,6 +168,38 @@ class BlockCompletion(TimeStampedModel, models.Model):
...
+
+        Return value:
+            dict[BlockKey] = float
+        """
+        course_block_completions = cls.objects.filter(
+            user=user,
+            course_key=course_key,
+        )
+        # will not return if <= 0.0
```
Comment (iloveagent57) [inline completion/models.py]: What does this comment mean?  Is it true?

File: completion/tests/test_models.py [hunk 7]
```diff
@@ -142,3 +142,69 @@ def test_submit_batch_completion_without_waffle(self):
...
+            )
+
+        models.BlockCompletion.objects.submit_completion(  # Wrong course
+            user=self.user,
+            course_key=self.other_course_key,
+            block_key=self.block_keys[4],
+            completion=0.75,
+        )
+
+    def test_get_course_completions(self):
```
Comment (iloveagent57) [inline completion/tests/test_models.py]: Can we have a test that checks that `get_course_completions()` returns an empty `dict` if the user has no completions for a course?

File: completion/tests/test_models.py [hunk 8]
```diff
@@ -142,3 +142,69 @@ def test_submit_batch_completion_without_waffle(self):
...
+        self.assertEqual(
+            models.BlockCompletion.get_course_completions(self.user, self.course_key),
+            {
+                self.block_keys[0]: 1.0,
+                self.block_keys[1]: 0.8,
+                self.block_keys[2]: 0.6,
+            },
+        )
+
+    def test_get_latest_block_completed(self):
```
Comment (iloveagent57) [inline completion/tests/test_models.py]: Can we have another test that `get_latest_block_completed()` returns None if there are no completions for the user in the course?

File: completion/waffle.py [hunk 9]
```diff
@@ -21,9 +23,67 @@
...
     """
     return WaffleSwitchNamespace(name=WAFFLE_NAMESPACE, log_prefix='completion: ')
+
+
+def waffle_flag():
+    """
+    Returns the namespaced, cached, audited Waffle flags dictionary for Completion.
+    """
+    namespace = WaffleFlagNamespace(name=WAFFLE_NAMESPACE, log_prefix=u'completion: ')
+    return {
```
Comment (iloveagent57) [inline completion/waffle.py]: Why does this return a `dict` - can it just return the `CourseWaffleFlag` object?

File: completion/waffle.py [hunk 10]
```diff
@@ -21,9 +23,67 @@
...
+    Exposes varia of visual progress feature.
+        ENABLE_COMPLETION_TRACKING, current_site.configuration, AND
+        enable_course_visual_progress OR enable_visual_progress
+
+    :return:
+
+        bool -> True if site/course/global enabled for visual progress tracking
+    """
+    if not waffle().is_enabled(ENABLE_COMPLETION_TRACKING):
+        return
```
Comment (iloveagent57) [inline completion/waffle.py]: (Warning, what I'm about to say is pedantic.)

The name of this method suggests that it will return a boolean value.  But all of these bare returns mean it will return either `None`, `True`, or `False`.  This doesn't matter much functionally, since `None` is False, but just to reduce any possible confusion, it would be better to explicitly return `False` in all of the places that you have a bare `return`.

File: completion/models.py (hunk 2 above)
Comment (yro) [reply completion/models.py]: That's fair.

File: completion/models.py (hunk 6 above)
Comment (yro) [reply completion/models.py]: @iloveagent57  It's true in practical terms (as of now on platform) -- but it's not true _by necessity_. I'll ditch the comment.

File: completion/models.py (hunk 1 above)
Comment (yro) [reply completion/models.py]: 💯

File: completion/tests/test_models.py (hunk 3 above)
Comment (yro) [reply completion/tests/test_models.py]: Sure!

File: completion/tests/test_models.py (hunk 4 above)
Comment (yro) [reply completion/tests/test_models.py]: I'll change it to `user_one` and `user_two`. Compromise!

File: completion/waffle.py (hunk 10 above)
Comment (yro) [reply completion/waffle.py]: Fair enough. I'm flying a little blind with this waffle stuff, so I'll make it a little prettier.

File: completion/waffle.py (hunk 9 above)
Comment (yro) [reply completion/waffle.py]: That's a fair question. I guess it's theoretically so that we can add other flags? I'll remove it and alter the receiving code in platform.

File: completion/tests/test_models.py (hunk 7 above)
Comment (yro) [reply completion/tests/test_models.py]: 👍

**Comment by yro** [comment]
@iloveagent57 I believe I have addressed your feedback.
